torch/testing/_internal/composite_compliance.py

```diff
@@ -0,0 +1,174 @@
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "shape of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+    if wrapper_tensor.dtype != elem.dtype:
+        raise RuntimeError(
+            "This operator is not CompositeImplicitAutograd compliant: the "
+            "dtype of the tensor was modified directly without "
+            "going through the PyTorch dispatcher.")
+
+# This is a bit fragile because this needs to be updated whenever a new view op
+# is added. We probably want some sort of introspection to tell that if a
+# torch.* op is a view or not.
+def is_view_fn(func):
+    return func.__name__ in {
+        'as_strided',
+        'detach',
+        'diagonal',
+        'expand',
+        'expand_as',
+        'movedim',
+        'narrow',
+        'permute',
+        'select',
+        'squeeze',
+        'transpose',
+        't',
+        'real',
+        'imag',
+        'view_as_real',
+        'view_as_complex',
+        'unflatten',
+        'unfold',
+        'unsqueeze',
+        'view',
+        'view_as',
+        'unbind',
+        'split',
+        'split_with_sizes',
+        'vsplit',
+        'hsplit',
+        'tensor_split',
+        'chunk',
+        'swapaxes',
+        'slice',
+        '_reshape_alias',
+        '_unsafe_view',
+        '_conj',
+    }
+
+def is_inplace_view_fn(func):
+    return func.__name__ in {
+        'squeeze_',
+        'unsqueeze_',
+        'transpose_',
+        't_',
+    }
+
+class CompositeCompliantTensor(torch.Tensor):
+    elem: torch.Tensor
+
+    __slots__ = ['elem']
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        # The storage of CompositeCompliantTensor should never be used directly
+        # by a CompositeImplicitAutograd operation; if the CompositeImplicitAutograd
+        # operator attempts to read from the storage without dispatching then it'll
+        # raise a RuntimeError due to it being a meta storage.
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls, elem.size(),
+            dtype=elem.dtype, layout=elem.layout,
+            device=elem.device, requires_grad=elem.requires_grad)
+        r.elem = elem
+        return r
+
+    def __repr__(self):
+        return f"CompositeCompliantTensor({self.elem})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e):
+            return e.elem if isinstance(e, CompositeCompliantTensor) else e
+
+        def wrap(e):
+            return CompositeCompliantTensor(e) if isinstance(e, torch.Tensor) else e
+
+        if func.__name__ in ('set_', 'resize_'):
+            raise RuntimeError(
+                f"{func.__name__} is not allowed to be called inside of "
+                f"CompositeImplicitAutograd operators.")
+
+        with no_dispatch():
```

Comment:
This is not needed right?

Reply:
This is needed to prevent infinite recursion if `func` is a factory function

Reply:
We should have a better story for this. In particular, this prevents any interaction between different subclasses because other subclass won't have their dispatch called anymore...

Reply:
btw I was just chatting with Anjali where the disable torch function was introducing subtle bugs in her implementation when subclass are involved.
For these, I think `super().__torch_function__(...)` is really the right way to go. I feel like we need a similar thing for torch_dispatch no?